### PR TITLE
Add google hashes

### PR DESCRIPTION
**Pull request by bbfsdev**
@genadyp please review.

### CODE REVIEW

File: lib/content_data/content_data.rb [hunk 1]
```diff
@@ -50,51 +63,33 @@ def initialize(other = nil)
...
 
     def clone_instances_info
-      clone_instances_info = {}
-      instances_info_enum = @instances_info.each_key
-      loop {
-        location = instances_info_enum.next rescue break
-        clone_instances_info[[location[0].clone, location[1].clone]] = @instances_info[location].clone
+      new_instances_info = GoogleHashSparseRubyToRuby.new
+      @instances_info.each { |key, value|
+        new_instances_info[key] = value
```
Comment (genadyp) [inline lib/content_data/content_data.rb:72]: Should  be used a clone method as it was in Yaron implementation?

File: lib/content_data/content_data.rb [hunk 2]
```diff
@@ -50,51 +63,33 @@ def initialize(other = nil)
...
-          # we use deep clone for location since map key is using shallow clone.
-          # we dont want references between new content data
-          # and orig object. This will help the GC dispose the orig object if not used any more.
-          instances_db_cloned[[location[0].clone,location[1].clone]] = inst_mod_times.clone
+      new_contents_info = GoogleHashSparseRubyToRuby.new
+      @contents_info.each { |checksum, content_info|
+        new_content_info = [content_info[0], GoogleHashSparseRubyToRuby.new, content_info[2]]
+        new_contents_info[checksum] = new_content_info
+        content_info[1].each { |location, stat|
+          new_content_info[1][location] = stat
```
Comment (genadyp) [inline lib/content_data/content_data.rb]: Should be used a clone method as it was in Yaron implementation?

File: lib/content_data/content_data.rb [hunk 3]
```diff
@@ -50,51 +63,33 @@ def initialize(other = nil)
...
 
     def clone_symlinks_info
-      symlinks_info_enum = @symlinks_info.each_key
-      cloned_symlinks = {}
-      loop {
-        symlink_key = symlinks_info_enum.next rescue break
-        cloned_symlinks[[symlink_key[0].clone, symlink_key[0].clone]] = @symlinks_info[symlink_key].clone
+      cloned_symlinks = GoogleHashSparseRubyToRuby.new
+      @symlinks_info.each{ |key, value|
+        cloned_symlinks[ket] = value
```
Comment (genadyp) [inline lib/content_data/content_data.rb]: Should be used a clone method as it was in Yaron implementation?

File: lib/content_data/content_data.rb [hunk 4]
```diff
@@ -115,17 +106,10 @@ def each_content(&block)
     # block is provided with: checksum, size, content modification time,
     #   instance modification time, server and file path
```
Comment (genadyp) [inline lib/content_data/content_data.rb]: comment should be updated: "instance index time" block parameter is missing

File: lib/content_data/content_data.rb [hunk 5]
```diff
@@ -171,15 +148,16 @@ def symlinks_size()
     end
 
     def checksum_instances_size(checksum)
```
Comment (genadyp) [inline lib/content_data/content_data.rb:152]: Method name is not informative.
Consider rename or add method description.

File: lib/content_data/content_data.rb (hunk 1 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb:72]: Value here is checksum, basic value, no need to clone it. There is a need to clone only arrays and hashes.
This code is not pretty, we really should use small classes for that.

File: lib/content_data/content_data.rb (hunk 2 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb]: This code is already updated with stat.clone().
Good catch.

File: lib/content_data/content_data.rb (hunk 3 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb]: Value here is target file, basic string, no need to clone.

File: lib/content_data/content_data.rb (hunk 4 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb]: Done.

File: lib/content_data/content_data.rb (hunk 5 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb:152]: Added description.

**Comment by bbfsdev** [comment]
Done.

File: lib/content_data/content_data.rb [hunk 6]
```diff
@@ -171,15 +148,16 @@ def symlinks_size()
...
+      return 0 unless @contents_info.key?(checksum)
       content_info = @contents_info[checksum]
-      return 0 if content_info.nil?
       content_info[1].length
     end
 
     def get_instance_mod_time(checksum, location)
+      return nil unless @contents_info.key?(checksum)
```
Comment (genadyp) [inline lib/content_data/content_data.rb:160]: to raise exception in this case?
Need a method description comment when nil returned or exception raised - it makes usage more secure..

File: lib/content_data/content_data.rb [hunk 7]
```diff
@@ -171,15 +148,16 @@ def symlinks_size()
...
       content_info = @contents_info[checksum]
-      return nil if content_info.nil?
       instances = content_info[1]
+      return nil
```
Comment (genadyp) [inline lib/content_data/content_data.rb]: this return makes following two lines unreachable

File: lib/content_data/content_data.rb [hunk 8]
```diff
@@ -171,15 +148,16 @@ def symlinks_size()
     end
 
     def checksum_instances_size(checksum)
+      return 0 unless @contents_info.key?(checksum)
       content_info = @contents_info[checksum]
-      return 0 if content_info.nil?
       content_info[1].length
     end
 
     def get_instance_mod_time(checksum, location)
```
Comment (genadyp) [inline lib/content_data/content_data.rb:159]: Need somewhere a general comment what is a time format we are working with.
It is not obvious that a return value of the method is an integer (Unix epoch time).

File: lib/content_data/content_data.rb [hunk 9]
```diff
@@ -256,40 +237,63 @@ def remove_instance(server, path)
...
-      loop {
-        checksum = contents_enum.next rescue break
-        instances =  @contents_info[checksum][1]
-        instances_enum = instances.each_key
-        loop {
-          location = instances_enum.next rescue break
+      @contents_info.each { |checksum, content_info|
+        locations_to_delete = []
+        content_info[1].each { |location, stats|
           if location[0] == server and location[1].scan(dir_to_remove).size > 0
```
Comment (genadyp) [inline lib/content_data/content_data.rb]: Looks like scan can return an unexpected results.
For example we'd like to remove directory "/dev/tools/"
then location /home/genadyp/dev/tools/ will be deleted:

> 1.9.3-p327 :022 >   dir = "/dev/tools/"
>  => "/dev/tools/" 
> 1.9.3-p327 :023 > path = "/home/genadyp/dev/tools/irb"
>  => "/home/genadyp/dev/tools/irb" 
> 1.9.3-p327 :024 > path.scan(dir)
>  => ["/dev/tools/"] 
> 1.9.3-p327 :025 > 

Is it OK?

File: lib/content_data/content_data.rb (hunk 7 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb]: This was the actual but why unify time did not work :)
Nice catch.

File: lib/content_data/content_data.rb (hunk 8 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb:159]: Done.

File: lib/content_data/content_data.rb (hunk 9 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb]: No, this is not ok, should be prefix.

File: lib/content_data/content_data.rb [hunk 10]
```diff
@@ -256,40 +240,68 @@ def remove_instance(server, path)
...
-        loop {
-          location = instances_enum.next rescue break
-          if location[0] == server and location[1].scan(dir_to_remove).size > 0
-            instances.delete(location)
-            @instances_info.delete(location)
+      @contents_info.each { |checksum, content_info|
+        locations_to_delete = []
+        content_info[1].each { |location, stats|
+          path_split = location[1].split(File::SEPARATOR)
+          dir_split = dir_to_remove.split(File::SEPARATOR)
```
Comment (genadyp) [inline lib/content_data/content_data.rb]: looks like it enough to run split on dir_to_remove only once, suggested to move it out of the loop.

File: lib/content_data/content_data.rb [hunk 11]
```diff
@@ -256,40 +240,68 @@ def remove_instance(server, path)
...
+          if path_split[0] != "" || dir_split[0] != ""
```
Comment (genadyp) [inline lib/content_data/content_data.rb]: looks like linux specific check and than can fall on windows.

File: lib/content_data/content_data.rb [hunk 12]
```diff
@@ -256,40 +240,68 @@ def remove_instance(server, path)
...
 
       # handle symlinks
-      symlinks_enum = @symlinks_info.each_key
-      loop {
-        symlink_key = symlinks_enum.next rescue break
-        if symlink_key[0] == server and symlink_key[1].scan(dir_to_remove).size > 0
-          @symlinks_info.delete(symlink_key)
+      @symlinks_info.each { |key, _|
+        keys_to_delete = []
+        if key[0] == server and key[1].scan(dir_to_remove).size > 0
```
Comment (genadyp) [inline lib/content_data/content_data.rb]: looks like this scan method should be treated as above

File: lib/content_data/content_data.rb [hunk 13]
```diff
@@ -256,40 +240,68 @@ def remove_instance(server, path)
...
         end
+        keys_to_delete.each { |key|
+          @symlinks_info.delete(key)
+        }
       }
     end
 
     def ==(other)
-      return nil if other.nil?  # for this case: content_data == nil
-      unique_id  == other.unique_id
```
Comment (genadyp) [inline lib/content_data/content_data.rb:286]: was found any problem with '==' in google_hash ?

File: lib/content_data/content_data.rb [hunk 14]
```diff
@@ -256,40 +240,68 @@ def remove_instance(server, path)
...
+        return false if other.instances_info[location] != checksum
+      }
+      return true
     end
 
     def remove_content(checksum)
       content_info = @contents_info[checksum]
       if content_info
-        content_info[1].each_key { |location|
+        content_info[1].each { |location, stats|
```
Comment (genadyp) [inline lib/content_data/content_data.rb]: stats parameter is unused - can be replaced by '_' in the block parameters

File: lib/content_data/content_data.rb [hunk 15]
```diff
@@ -299,25 +311,24 @@ def remove_content(checksum)
...
       }
       each_instance { |checksum, size, content_mod_time, instance_mod_time, server, path|
-        instances_str <<  "%s,%d,%s,%s,%d\n" % [checksum, size, server, path, instance_mod_time]
+        instances <<  "%s,%d,%s,%s,%d" % [checksum, size, server, path, instance_mod_time]
       }
       return_str << "%d\n" % [@contents_info.length]
-      return_str << contents_str
-      return_str << "%d\n" % [@instances_info.length]
-      return_str << instances_str
+      return_str << contents.sort.join("\n")
```
Comment (genadyp) [inline lib/content_data/content_data.rb]: suggest to check how many time takes sort on a large array

File: lib/content_data/content_data.rb (hunk 13 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb:286]: No, this comparison is bad, it is based on internal unique_id which is wrong here, what if we have to identical copies then the unique_id will be different. That is not a way to do comparison.

File: lib/content_data/content_data.rb (hunk 10 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb]: Done.

File: lib/content_data/content_data.rb (hunk 11 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb]: Removed this check.

File: lib/content_data/content_data.rb (hunk 12 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb]: Done. Added tests to symlinks - there was none!!!

File: lib/content_data/content_data.rb (hunk 14 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb]: Marked as unused_stats

File: lib/content_data/content_data.rb (hunk 15 above)
Comment (bbfsdev) [inline lib/content_data/content_data.rb]: This is for debug only purposes, we don't use print in other situations.
Why do we need to check how many time takes sort? Should be n*log(n).
Anyway we should not do this unless we have speed problems.

**Comment by bbfsdev** [comment]
Done, please approve to merge...

File: spec/file_utils/time_modification_spec.rb [hunk 16]
```diff
@@ -41,6 +41,7 @@ module Spec
         # must preced Log.init, otherwise log containing default values will be created
         Params['log_write_to_file'] = false
         Params['log_write_to_console'] = false
+        Params['log_debug_level'] = 1
```
Comment (genadyp) [inline spec/file_utils/time_modification_spec.rb:44]: Do we actually need to output debug messages?

File: spec/file_utils/time_modification_spec.rb [hunk 17]
```diff
@@ -127,6 +128,7 @@ module Spec
         }
 
         # checking that files were actually modified
+        # TODO(kolman): Decide whether files should be updated or not and change the implementation.
```
Comment (genadyp) [inline spec/file_utils/time_modification_spec.rb:131]: Consider to open issue about this TODO

**Comment by genadyp** [comment]
@kolmanv 
Review is finished.
There was few minor additional comments.
You can merge.
Thank you :)

**Pull request merged**